Description: all: add "Reported OS" option
Author: Dmitrii Sutiagin <f3flight@gmail.com>
Origin: upstream, https://gitlab.gnome.org/GNOME/NetworkManager-openconnect/commit/9fbe2817da437f473cf46205e19c38d273adc600
Bug: https://bugzilla.gnome.org/show_bug.cgi?id=766239
Bug: https://bugzilla.gnome.org/show_bug.cgi?id=782016
Bug: https://bugzilla.gnome.org/show_bug.cgi?id=783846
Last-Update: 2017-05-01
---
This patch header follows DEP-3: http://dep.debian.net/deps/dep3/
--- a/auth-dialog/main.c
+++ b/auth-dialog/main.c
@@ -963,6 +963,7 @@
 	char *csd;
 	char *sslkey, *cert;
 	char *csd_wrapper;
+	char *reported_os;
 	char *pem_passphrase_fsid;
 	char *cafile;
 	char *token_mode;
@@ -1033,6 +1034,10 @@
 		openconnect_setup_csd(vpninfo, getuid(), 1, OC3DUP (csd_wrapper));
 	}
 
+	reported_os = g_hash_table_lookup (options, NM_OPENCONNECT_KEY_REPORTED_OS);
+	if (reported_os && reported_os[0])
+		openconnect_set_reported_os(vpninfo, reported_os);
+
 	proxy = g_hash_table_lookup (options, NM_OPENCONNECT_KEY_PROXY);
 	if (proxy && proxy[0] && openconnect_set_http_proxy(vpninfo, OC3DUP (proxy)))
 		return -EINVAL;
--- a/properties/nm-openconnect-dialog.ui
+++ b/properties/nm-openconnect-dialog.ui
@@ -109,7 +109,7 @@
                             <property name="editable">True</property>
                             <property name="visibility">True</property>
                             <property name="max_length">0</property>
-                            <property name="text" translatable="yes"/>
+                            <property name="text"/>
                             <property name="has_frame">True</property>
                             <property name="invisible_char">&#x2022;</property>
                             <property name="activates_default">False</property>
@@ -142,7 +142,7 @@
                             <property name="editable">True</property>
                             <property name="visibility">True</property>
                             <property name="max_length">0</property>
-                            <property name="text" translatable="yes"/>
+                            <property name="text"/>
                             <property name="has_frame">True</property>
                             <property name="invisible_char">&#x2022;</property>
                             <property name="activates_default">False</property>
@@ -361,7 +361,7 @@
                             <property name="editable">True</property>
                             <property name="visibility">True</property>
                             <property name="max_length">0</property>
-                            <property name="text" translatable="yes"/>
+                            <property name="text"/>
                             <property name="has_frame">True</property>
                             <property name="invisible_char">&#x2022;</property>
                             <property name="activates_default">False</property>
@@ -393,6 +393,90 @@
                         <property name="ellipsize">PANGO_ELLIPSIZE_NONE</property>
                         <property name="width_chars">-1</property>
                         <property name="single_line_mode">False</property>
+                        <property name="angle">0</property>
+                      </object>
+                      <packing>
+                        <property name="left_attach">0</property>
+                        <property name="right_attach">1</property>
+                        <property name="top_attach">0</property>
+                        <property name="bottom_attach">1</property>
+                        <property name="y_options"/>
+                      </packing>
+                    </child>
+                  </object>
+                </child>
+              </object>
+            </child>
+            <child>
+              <object class="GtkAlignment" id="alignment18">
+                <property name="visible">True</property>
+                <property name="xalign">0.5</property>
+                <property name="yalign">0.5</property>
+                <property name="xscale">1</property>
+                <property name="yscale">1</property>
+                <property name="top_padding">0</property>
+                <property name="bottom_padding">0</property>
+                <property name="left_padding">12</property>
+                <property name="right_padding">0</property>
+                <child>
+                  <object class="GtkTable" id="table5">
+                    <property name="visible">True</property>
+                    <property name="n_rows">1</property>
+                    <property name="n_columns">2</property>
+                    <property name="homogeneous">False</property>
+                    <property name="row_spacing">6</property>
+                    <property name="column_spacing">6</property>
+                    <child>
+                      <object class="GtkAlignment" id="alignment19">
+                        <property name="visible">True</property>
+                        <property name="xalign">1</property>
+                        <property name="yalign">0.5</property>
+                        <property name="xscale">0</property>
+                        <property name="yscale">1</property>
+                        <property name="top_padding">0</property>
+                        <property name="bottom_padding">0</property>
+                        <property name="left_padding">0</property>
+                        <property name="right_padding">0</property>
+                        <child>
+                          <object class="GtkEntry" id="reported_os">
+                            <property name="visible">True</property>
+                            <property name="can_focus">True</property>
+                            <property name="editable">True</property>
+                            <property name="visibility">True</property>
+                            <property name="max_length">0</property>
+                            <property name="text"/>
+                            <property name="has_frame">True</property>
+                            <property name="invisible_char">&#x2022;</property>
+                            <property name="activates_default">False</property>
+                          </object>
+                        </child>
+                      </object>
+                      <packing>
+                        <property name="left_attach">1</property>
+                        <property name="right_attach">2</property>
+                        <property name="top_attach">0</property>
+                        <property name="bottom_attach">1</property>
+                        <property name="y_options"/>
+                      </packing>
+                    </child>
+                    <child>
+                      <object class="GtkLabel" id="label1">
+                        <property name="visible">True</property>
+                        <property name="label" translatable="yes">Reported OS:</property>
+                        <property name="tooltip_text" translatable="yes">Common values: linux linux-64 win mac-intel android apple-ios</property>
+                        <property name="use_underline">True</property>
+                        <property name="use_markup">False</property>
+                        <property name="justify">GTK_JUSTIFY_LEFT</property>
+                        <property name="wrap">False</property>
+                        <property name="selectable">False</property>
+                        <property name="xalign">0</property>
+                        <property name="yalign">0.5</property>
+                        <property name="xpad">0</property>
+                        <property name="ypad">0</property>
+                        <property name="mnemonic_widget">reported_os</property>
+                        <property name="ellipsize">PANGO_ELLIPSIZE_NONE</property>
+                        <property name="width_chars">-1</property>
+                        <property name="single_line_mode">False</property>
                         <property name="angle">0</property>
                       </object>
                       <packing>
--- a/properties/nm-openconnect-editor-plugin.c
+++ b/properties/nm-openconnect-editor-plugin.c
@@ -179,6 +179,11 @@
 	if (buf)
 		nm_setting_vpn_add_data_item (s_vpn, NM_OPENCONNECT_KEY_CSD_WRAPPER, buf);
 
+	/* Reported OS */
+	buf = g_key_file_get_string (keyfile, "openconnect", "ReportedOS", NULL);
+	if (buf)
+		nm_setting_vpn_add_data_item (s_vpn, NM_OPENCONNECT_KEY_REPORTED_OS, buf);
+
 	/* User Certificate */
 	buf = g_key_file_get_string (keyfile, "openconnect", "UserCertificate", NULL);
 	if (buf && strcmp(buf, "(null)"))
@@ -222,6 +227,7 @@
 	const char *proxy = NULL;
 	gboolean csd_enable = FALSE;
 	const char *csd_wrapper = NULL;
+	const char *reported_os = NULL;
 	const char *usercert = NULL;
 	const char *privkey = NULL;
 	gboolean pem_passphrase_fsid = FALSE;
@@ -274,6 +280,10 @@
 	if (value && strlen (value))
 		csd_wrapper = value;
 
+	value = nm_setting_vpn_get_data_item (s_vpn, NM_OPENCONNECT_KEY_REPORTED_OS);
+	if (value && strlen (value))
+		reported_os = value;
+
 	value = nm_setting_vpn_get_data_item (s_vpn, NM_OPENCONNECT_KEY_USERCERT);
 	if (value && strlen (value))
 		usercert = value;
@@ -308,6 +318,7 @@
 		 "Proxy=%s\n"
 		 "CSDEnable=%s\n"
 		 "CSDWrapper=%s\n"
+		 "ReportedOS=%s\n"
 		 "UserCertificate=%s\n"
 		 "PrivateKey=%s\n"
 		 "FSID=%s\n"
@@ -320,6 +331,7 @@
 		 /* Proxy */                 proxy ? proxy : "",
 		 /* Cisco Secure Desktop */  csd_enable ? "1" : "0",
 		 /* CSD Wrapper Script */    csd_wrapper ? csd_wrapper : "",
+		 /* Reported OS */           reported_os ? reported_os : "",
 		 /* User Certificate */      usercert ? usercert : "",
 		 /* Private Key */           privkey ? privkey : "",
 		 /* FSID */                  pem_passphrase_fsid ? "1" : "0",
--- a/properties/nm-openconnect-editor.c
+++ b/properties/nm-openconnect-editor.c
@@ -346,6 +346,16 @@
 	}
 	g_signal_connect (G_OBJECT (widget), "changed", G_CALLBACK (stuff_changed_cb), self);
 
+	widget = GTK_WIDGET (gtk_builder_get_object (priv->builder, "reported_os"));
+	if (!widget)
+		return FALSE;
+	if (s_vpn) {
+		value = nm_setting_vpn_get_data_item (s_vpn, NM_OPENCONNECT_KEY_REPORTED_OS);
+		if (value)
+			gtk_entry_set_text (GTK_ENTRY (widget), value);
+	}
+	g_signal_connect (G_OBJECT (widget), "changed", G_CALLBACK (stuff_changed_cb), self);
+
 	if (init_token_ui (self, priv, s_vpn) == FALSE)
 		return FALSE;
 
@@ -425,6 +435,11 @@
 	if (str && strlen (str))
 		nm_setting_vpn_add_data_item (s_vpn, NM_OPENCONNECT_KEY_CSD_WRAPPER, str);
 
+	widget = GTK_WIDGET (gtk_builder_get_object (priv->builder, "reported_os"));
+	str = (char *) gtk_entry_get_text (GTK_ENTRY (widget));
+	if (str && strlen (str))
+		nm_setting_vpn_add_data_item (s_vpn, NM_OPENCONNECT_KEY_REPORTED_OS, str);
+
 	widget = GTK_WIDGET (gtk_builder_get_object (priv->builder, "token_mode"));
 	model = gtk_combo_box_get_model (GTK_COMBO_BOX (widget));
 	if (gtk_combo_box_get_active_iter (GTK_COMBO_BOX (widget), &iter)) {
--- a/shared/nm-service-defines.h
+++ b/shared/nm-service-defines.h
@@ -46,5 +46,6 @@
 #define NM_OPENCONNECT_KEY_CSD_WRAPPER "csd_wrapper"
 #define NM_OPENCONNECT_KEY_TOKEN_MODE "stoken_source"
 #define NM_OPENCONNECT_KEY_TOKEN_SECRET "stoken_string"
+#define NM_OPENCONNECT_KEY_REPORTED_OS "reported_os"
 
 #endif /* __NM_SERVICE_DEFINES_H__ */
--- a/src/nm-openconnect-service.c
+++ b/src/nm-openconnect-service.c
@@ -92,6 +92,7 @@
 	{ NM_OPENCONNECT_KEY_CSD_WRAPPER, G_TYPE_STRING, 0, 0 },
 	{ NM_OPENCONNECT_KEY_TOKEN_MODE,  G_TYPE_STRING, 0, 0 },
 	{ NM_OPENCONNECT_KEY_TOKEN_SECRET, G_TYPE_STRING, 0, 0 },
+	{ NM_OPENCONNECT_KEY_REPORTED_OS, G_TYPE_STRING, 0, 0 },
 	{ NULL,                           G_TYPE_NONE, 0, 0 }
 };
 
